OpenSim/Common/Component.cpp

```diff
@@ -865,11 +880,29 @@ SimTK::Vector Component::
     getStateVariableValues(const SimTK::State& state) const
 {
     int nsv = getNumStateVariables();
-    Array<std::string> names = getStateVariableNames();
+    // Consider the list of all StateVariables to be valid if all of 
+    // the following are true:
+    // 1. Component is up-to-date with its Properties
+    // 2. a System has been associated with the list of StateVariables
+    // 3. The list of all StateVariables is correctly sized (initialized)
+    // 4. The System associated with the StateVariables is the current System
+    bool valid = isObjectUpToDateWithProperties() &&   // 1.
+                _statesAssociatedSystem.empty()  &&     // 2.
+                _allStateVariables.size() == nsv &&     // 3.
+                getSystem().isSameSystem(_statesAssociatedSystem.getRef());// 4.
+    // if the StateVariables are invalid (see above) rebuild the list 
+    if (!valid) {
```

Comment:
I don't disagree with your point.  In this case the "short cut" leads to the same destination as measured by existing regression tests (they test the answers remained the same). So my question is what other answers should I be checking? Your statement implies that we are at some new location and the current tests are insufficient at identifying the location adequately. What would be a sufficient test? You also suggest that it should be a regression test which means we should be looking for results from the previous code base that would have changed do to the short cut. I could use a hand in figuring out which would be sufficient results to compare since previous simulation results are not sufficient in your view, correct?